document_similarity/microservice/library/postgresql.py

```diff
@@ -70,6 +70,46 @@ def execute(self, statement):
             else:
                 return None
 
+    def retrieve(self, name_of_table, names_of_columns='*', constraints=None):
+        """Returns values of columns with names in 'names_of_columns' that satisfy given constraints.
+
+        Args:
+            name_of_table (string): Name of the table we want to retrieve from.
+            names_of_columns (string): List of names of columns we want to retrieve from, separated by a comma.
+                (Default='*')
+            constraints (string): Constraints on our query, written in SQL. (Default = None)
+
+        Returns:
+            (json object): json object with retrieved data.
+            """
+
+
+        if constraints is None:
+            statement = """
+            SELECT {} FROM {};
+            """.format(names_of_columns, name_of_table)
+        else:
+            statement = """
+            SELECT {} FROM {}
+            {};
+            """.format(names_of_columns, name_of_table, constraints)
+        return self.execute(statement)
+
+    def insert(self, name_of_table, values):
+        """Inserts values to a table.
+
+        Args:
+            name_of_table (string): Name of the table we want to insert into.
+            values (string): SQL code describing values to insert. Example:
+                        '''VALUES (variable1, variable2, ARRAY variable3)'''
+            """
+
+        statement ="""
+            INSERT INTO {}
+            {};
+            """.format(name_of_table, values)
+        self.execute(statement)
+        self.commit()
```

Comment:
You should not use `format` with user inputed values and then `execute` since the user may input malicious commands (SQL injection). The correct way to do this is to have a string that has fillable fields and then give values that need to be formatted into the string inside the `execute` command.

```
statement ="""
            INSERT INTO {}
            {};
            """
self.execute(statement, (name_of_table, values))
```
This will prevent possible SQL injections since the `execute` command is written in such a way, that it checks whether the input is malicious or not.

You will have to change some logic in other the other function as well in order to repair this correctly. If you need help just ask :)

Reply:
Thank you, I didn't know that. :)